// react-native-lucky-wheel/lib/commonjs/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = exports.TextAngles = exports.GestureTypes = exports.EasingTypes = void 0;

var _components = require("./components");

let GestureTypes;
exports.GestureTypes = GestureTypes;

(function (GestureTypes) {
  GestureTypes["CLOCKWISE"] = "clockwise";
  GestureTypes["ANTI_CLOCKWISE"] = "anti-clockwise";
  GestureTypes["MULTIDIRECTIONAL"] = "multidirectional";
})(GestureTypes || (exports.GestureTypes = GestureTypes = {}));

let EasingTypes;
exports.EasingTypes = EasingTypes;

(function (EasingTypes) {
  EasingTypes["OUT"] = "out";
  EasingTypes["IN_OUT"] = "in_out";
})(EasingTypes || (exports.EasingTypes = EasingTypes = {}));

let TextAngles;
exports.TextAngles = TextAngles;

(function (TextAngles) {
  TextAngles["VERTICAL"] = "vertical";
  TextAngles["HORIZONTAL"] = "horizontal";
})(TextAngles || (exports.TextAngles = TextAngles = {}));

var _default = _components.LuckyWheel;
exports.default = _default;
