// react-native-fast-confetti/lib/module/hooks/useConfettiLogic.js
"use strict";

import { useDerivedValue } from 'react-native-reanimated';
import { useTexture, Group, RoundedRect, rect } from '@shopify/react-native-skia';
import { jsx as _jsx } from "react/jsx-runtime";
export const useConfettiLogic = ({
  sizeVariations,
  colors,
  boxes
}) => {
  const maxWidth = Math.max(...sizeVariations.map(size => size.width));
  const maxHeight = Math.max(...sizeVariations.map(size => size.height));
  const texture = useTexture(/*#__PURE__*/_jsx(Group, {
    children: colors.map((color, colorIndex) => {
      return sizeVariations.map((size, sizeIndex) => {
        return /*#__PURE__*/_jsx(RoundedRect, {
          x: sizeIndex * maxWidth,
          y: colorIndex * maxHeight,
          width: size.width,
          height: size.height,
          r: size.radius,
          color: color
        }, `${colorIndex}-${sizeIndex}`);
      });
    })
  }), {
    width: maxWidth * sizeVariations.length,
    height: maxHeight * colors.length
  });
  const sprites = useDerivedValue(() => {
    return boxes.value.map(box => {
      const colorIndex = box.colorIndex;
      const sizeIndex = box.sizeIndex;
      const size = sizeVariations[sizeIndex];
      return rect(sizeIndex * maxWidth, colorIndex * maxHeight, size.width, size.height);
    });
  });
  return {
    texture,
    sprites
  };
};
